examples/proxy/proxy.js

```diff
@@ -0,0 +1,237 @@
+const net = require('net')
+const { pipeline } = require('stream')
+
+module.exports = async function proxy (ctx, client) {
+  const { req, socket } = ctx
+
+  // TODO: requestTimeout?
+
+  if (socket) {
+    const handler = new WSHandler(ctx)
+    client.dispatch({
+      method: req.method,
+      path: req.originalUrl || req.url,
+      headers: getRequestHeaders(ctx),
+      upgrade: 'Websocket'
+    }, handler)
+    return handler.promise
+  } else {
+    const handler = new HTTPHandler(ctx)
+    client.dispatch({
+      method: req.method,
+      path: req.originalUrl || req.url,
+      headers: getRequestHeaders(ctx),
+      body: req
+    }, handler)
+    return handler.promise
+  }
+}
+
+class WSHandler {
+  constructor ({ socket, head }) {
+    setupSocket(socket)
+
+    this.socket = socket
+    this.head = head
+    this.promise = new Promise((resolve, reject) => {
+      this.callback = err => err ? reject(err) : resolve()
+    })
+  }
+
+  onConnect (abort) {
+    if (this.socket.destroyed) {
+      abort()
+    } else {
+      this.socket.on('close', abort)
+    }
+  }
+
+  onUpgrade (statusCode, headers, socket) {
+    if (this.head && this.head.length) {
+      socket.unshift(this.head)
+    }
+
+    headers = getResponseHeaders(headers)
+
+    setupSocket(socket)
+
+    let head = 'HTTP/1.1 101 Switching Protocols\r\nconnection: upgrade\r\nupgrade: websocket'
+
+    for (let n = 0; n < headers.length; n += 2) {
+      const key = headers[n + 0]
+      const val = headers[n + 1]
+
+      if (!Array.isArray(val)) {
+        head += `\r\n${key}: ${val}`
+      } else {
+        for (let i = 0; i < val.length; i++) {
+          head += `\r\n${key}: ${val[i]}`
+        }
+      }
+    }
+    head += '\r\n\r\n'
+
+    this.socket.write(head)
+
+    pipeline(socket, this.socket, socket, this.callback)
+  }
+
+  onError (err) {
+    this.callback(err)
+  }
+}
+
+class HTTPHandler {
+  constructor ({ req, res }) {
+    this.req = req
+    this.res = res
+    this.promise = new Promise((resolve, reject) => {
+      this.callback = err => err ? reject(err) : resolve()
+    })
+  }
+
+  onConnect (abort) {
+    if (this.req.aborted) {
+      abort()
+    } else {
+      this.res.on('close', abort)
+    }
+  }
+
+  onHeaders (statusCode, headers, resume) {
+    if (statusCode < 200) {
+      return
+    }
+
+    this.res.on('drain', resume)
+
+    headers = getResponseHeaders(headers)
+
+    // TODO (perf): res.writeHead should support Array and/or string.
```

Comment:
@mcollina: It would be nice if core `writeHead` support the key/value array directly without having to convert it into an object.